### PR TITLE
JitCache: Store the JIT blocks in the std::map.

### PR DESCRIPTION
**Pull request by degasus**
No need to provide a fixed size std::array. No need to mark blocks as invalid.

This PR also fixes block re-linking if the second one got invalidated.

It also provides a simple flag within the Jit64 dispatcher to skip the assembly one.

And it support the very rare case of caching two blocks for the same PC but different MSR. Mostly because it is simpler, not because it is required for performance.

### CODE REVIEW

**Comment by JMC47** [comment]
I can't see any big difference in Rogue Squadron 2 regarding performance with this.  Haven't done any super refined testing on it though.

**Comment by degasus** [comment]
@JMC47 yeah, it was just a tiny regression. Just wanted to be sure through. So thanks.

File: Source/Core/Core/PowerPC/JitCommon/JitCache.cpp [hunk 1]
```diff
@@ -94,56 +88,53 @@ void JitBaseBlockCache::SchedulateClearCacheThreadSafe()
...
+    else
+      iter.first++;
   }
-  start_block_map[block.physicalAddress] = &block;
-  FastLookupEntryForAddress(block.effectiveAddress) = &block;
+}
+
+void JitBaseBlockCache::FinalizeBlock(JitBlock& block, bool block_link, const u8* code_ptr)
+{
+  size_t icache = FastLookupEntryForAddress(block.effectiveAddress);
```
Comment (BhaaLseN) [inline Source/Core/Core/PowerPC/JitCommon/JitCache.cpp]: Please rename this variable, the next line only differs by casing and makes it very odd to read

File: Source/Core/Core/PowerPC/JitCommon/JitCache.cpp [hunk 2]
```diff
@@ -347,12 +330,19 @@ void JitBaseBlockCache::MoveBlockIntoFastCache(u32 addr, u32 msr)
   }
   else
   {
-    FastLookupEntryForAddress(addr) = block;
+    // Drop old icache entry
+    if (iCache[block->in_icache] == block)
+      iCache[block->in_icache] = nullptr;
+
+    // And create a new one
+    size_t icache = FastLookupEntryForAddress(addr);
```
Comment (BhaaLseN) [inline Source/Core/Core/PowerPC/JitCommon/JitCache.cpp]: Same here

File: Source/Core/Core/PowerPC/JitCommon/JitCache.cpp [hunk 3]
```diff
@@ -347,12 +330,19 @@ void JitBaseBlockCache::MoveBlockIntoFastCache(u32 addr, u32 msr)
...
+    iCache[icache] = block;
+    block->in_icache = icache;
     LinkBlock(*block);
   }
 }
 
-JitBlock*& JitBaseBlockCache::FastLookupEntryForAddress(u32 address)
+size_t JitBaseBlockCache::FastLookupEntryForAddress(u32 address)
```
Comment (BhaaLseN) [inline Source/Core/Core/PowerPC/JitCommon/JitCache.cpp]: The name of this method sounds a little odd. The old one actually did a lookup and returned an item, but this one returns...an index, key, offset, something like that?

**Pull request merged**